arch/lkl/kernel/syscalls.c

```diff
@@ -33,400 +32,136 @@ syscall_handler_t syscall_table[__NR_syscalls] = {
-	void *mutex, *completion;
-	int irq;
-	/* to be accessed from Linux context only */
-	wait_queue_head_t wqh;
-	struct list_head list;
-	bool stop;
-	struct completion stopped;
-} default_syscall_thread_data;
-
-static LIST_HEAD(syscall_threads);
-
-static struct syscall *dequeue_syscall(struct syscall_thread_data *data)
-{
-
-	return (struct syscall *)__sync_fetch_and_and((long *)&data->s, 0);
-}
-
-static long run_syscall(struct syscall *s)
+static long run_syscall(long no, long *params)
 {
 	long ret;
 
-	if (s->no < 0 || s->no >= __NR_syscalls)
-		ret = -ENOSYS;
-	else {
-		ret = syscall_table[s->no](s->params[0], s->params[1],
-					   s->params[2], s->params[3],
-					   s->params[4], s->params[5]);
-	}
-	s->ret = ret;
+	if (no < 0 || no >= __NR_syscalls)
+		return -ENOSYS;
+
+	ret = syscall_table[no](params[0], params[1], params[2], params[3],
+				params[4], params[5]);
 
 	task_work_run();
 
 	return ret;
 }
 
-static irqreturn_t syscall_irq_handler(int irq, void *dev_id)
-{
-	struct syscall_thread_data *data = (struct syscall_thread_data *)dev_id;
 
-	wake_up(&data->wqh);
+#define CLONE_FLAGS (CLONE_VM | CLONE_FS | CLONE_FILES | CLONE_THREAD |	\
+		     CLONE_SIGHAND | SIGCHLD)
 
-	return IRQ_HANDLED;
-}
+static int host_task_id;
+static struct task_struct *host0;
 
-static void cleanup_syscall_threads(void);
-
-int syscall_thread(void *_data)
+static int new_host_task(struct task_struct **task)
 {
-	struct syscall_thread_data *data;
-	struct syscall *s;
-	int ret;
-	static int count;
-
-	data = (struct syscall_thread_data *)_data;
-	init_waitqueue_head(&data->wqh);
-	list_add(&data->list, &syscall_threads);
-	init_completion(&data->stopped);
-
-	snprintf(current->comm, sizeof(current->comm), "ksyscalld%d", count++);
-
-	data->irq = lkl_get_free_irq("syscall");
-	if (data->irq < 0) {
-		pr_err("lkl: %s: failed to allocate irq: %d\n", __func__,
-		       data->irq);
-		return data->irq;
-	}
-
-	ret = request_irq(data->irq, syscall_irq_handler, 0, current->comm,
-			  data);
-	if (ret) {
-		pr_err("lkl: %s: failed to request irq %d: %d\n", __func__,
-		       data->irq, ret);
-		lkl_put_irq(data->irq, "syscall");
-		data->irq = -1;
-		return ret;
-	}
-
-	pr_info("lkl: syscall thread %s initialized (irq%d)\n", current->comm,
-		data->irq);
-
-	/* system call thread is ready */
-	lkl_ops->sem_up(data->completion);
-
-	while (1) {
-		wait_event(data->wqh,
-			   (s = dequeue_syscall(data)) != NULL || data->stop);
-
-		if (data->stop || s->no == __NR_reboot)
-			break;
-
-		run_syscall(s);
+	pid_t pid;
 
-		lkl_ops->sem_up(data->completion);
-	}
+	switch_to_host_task(host0);
 
-	if (data == &default_syscall_thread_data)
-		cleanup_syscall_threads();
+	pid = kernel_thread(NULL, NULL, CLONE_FLAGS);
+	if (pid < 0)
+		return pid;
 
-	pr_info("lkl: exiting syscall thread %s\n", current->comm);
+	rcu_read_lock();
+	*task = find_task_by_pid_ns(pid, &init_pid_ns);
+	rcu_read_unlock();
 
-	list_del(&data->list);
+	get_task_struct(*task);
```

Comment:
To avoid races when terminating a task, in __switch_to. Say we end a host task with del_host_task. We will call do_exit which eventually calls schedule and __switch_to. The first thing we do is to release the semaphore which will run finish_task_switch which in turn will issue put_task_struct which will free the thread_info structure. This can happen before we have a chance to test the flags in the thread that shutsdown.

Yes, we need a put_struct before thread_exit() in __switch_to but it got lost in the iterations :) Or perhaps it is simpler to make a copy of the _prev flags on stack.

Good catch, thanks !